miner/src/block_assembler.rs

```diff
@@ -470,43 +468,48 @@ impl<CS: ChainStore + 'static> BlockAssembler<CS> {
         Ok(output)
     }
 
-    /// A block B1 is considered to be the uncle of
-    /// another block B2 if all of the following conditions are met:
-    /// (1) they are in the same epoch, sharing the same difficulty;
-    /// (2) height(B2) > height(B1);
-    /// (3) B2 is the first block in its chain to refer to B1
+    // A block B1 is considered to be the uncle of another block B2 if all of the following conditions are met:
+    // (1) they are in the same epoch, sharing the same difficulty;
+    // (2) height(B2) > height(B1);
+    // (3) B1's parent is either B2's ancestor or embedded in B2 or its ancestors as an uncle;
+    // and (4) B2 is the first block in its chain to refer to B1.
     fn prepare_uncles(
         &self,
         candidate_number: BlockNumber,
         current_epoch_ext: &EpochExt,
-        candidate_uncles: &mut LruCache<H256, Arc<Block>>,
-    ) -> FnvHashSet<UncleBlock> {
+        candidate_uncles: &mut CandidateUncles,
+    ) -> Vec<UncleBlock> {
         let store = self.shared.store();
         let epoch_number = current_epoch_ext.number();
         let max_uncles_num = self.shared.consensus().max_uncles_num();
-        let mut uncles = FnvHashSet::with_capacity_and_hasher(max_uncles_num, Default::default());
+        let mut uncles: Vec<UncleBlock> = Vec::with_capacity(max_uncles_num);
```

Comment:
why not use HashSet?

Reply:
Need a sorted map,  keep the highest candidate uncle